tests/http/testenv/ports.py

```diff
@@ -24,24 +24,75 @@
 #
 ###########################################################################
 #
+import json
 import logging
+import os
+import re
 import socket
-from typing import Dict
+from json import JSONEncoder
+from typing import Dict, List
+
+import pytest
+from filelock import FileLock
 
 log = logging.getLogger(__name__)
 
 
-def alloc_ports(port_specs: Dict[str, int]) -> Dict[str, int]:
-    ports = {}
+def alloc_port_sets(n: int, port_specs: Dict[str, int]) -> List[Dict[str, int]]:
+    port_sets = []
     socks = []
-    for name, ptype in port_specs.items():
-        try:
-            s = socket.socket(type=ptype)
-            s.bind(('', 0))
-            ports[name] = s.getsockname()[1]
-            socks.append(s)
-        except Exception as e:
-            raise e
+    for _ in range(n):
+        ports = {}
+        for name, ptype in port_specs.items():
+            try:
+                s = socket.socket(type=ptype)
+                s.bind(('', 0))
+                ports[name] = s.getsockname()[1]
+                socks.append(s)
+            except Exception as e:
+                raise e
+        port_sets.append(ports)
     for s in socks:
         s.close()
-    return ports
+    return port_sets
+
+
+def alloc_ports(config: pytest.Config,
+                gen_dir,
+                testrun_uid,
+                worker_id : str,
+                port_specs: Dict[str, int]) -> Dict[str, int]:
+    if "PYTEST_XDIST_WORKER_COUNT" in os.environ:
+        nworkers = int(os.environ["PYTEST_XDIST_WORKER_COUNT"])
+    else:
+        nworkers = 1
+    m = re.match(r'\D+(\d+)', worker_id)
+    idx = int(m.group(1)) if m else 0
+    assert idx < nworkers
+    # several worker processes that need this set of ports. Make
+    # a global locked allocation for all of them, store in a file
+    #
+    # `testrun_uid` is the pytest-xdist generated id of this run.
+    lock_file = os.path.join(gen_dir, 'allocated.ports.lock')
```

Comment:
Simplify by appending `.lock` to port_file.

Reply:
Ack.